### PR TITLE
Changes for Complete coroutine.

### PR DESCRIPTION
**Pull request by favalos**
WIP for #166.

@guergabo I did the changes for this issue.

I marked this as Draft because I am getting a couple of test failures, if you have any advice it would be great. 

Any comment or suggestions are welcome.

Thanks!

### CODE REVIEW

**Comment by dfarr** [comment]
Thank you for the PR again @favalos!

I think the reason you are seeing test failures is because you re-ordered the states [here](https://github.com/resonatehq/resonate/pull/218/files#diff-46480e4c0b744a50df1e74167003c3579eda8ee719e1a47b7c31dcc1ba69c905R46-R47). So now the constants map to:
```go
const (
	Pending State = 1 << iota
	Resolved // 2
	Rejected // 4
	Canceled // 8
	Timedout // 16
	Invalid  // 32
)
```

I like this ordering better but we will need to change the test cases to match for the `Timedout` case.

We will also need to remove Invalid from this enum and guard against ever getting to this point. Instead I recommend changing the type of State in the `CompletePromiseBody` to `promise.State` that way we can take advantage of the [UnmarshalJSON function](https://github.com/resonatehq/resonate/blob/0773faba49a50a2ce85d1e8b0029909f8783f364/pkg/promise/promise.go#L70) we already have and the validation would become

```
!body.State.In(promise.Resolved|promise.Rejected|promise.Canceled)
```

**Comment by favalos** [comment]
Hey @dfarr,

Thanks for the hint, I re-ordered the Enum and that did the trick.

I fixed another couple of errors that I found and did the changes related to the recommendation mentioned above.

Thank you!

**Comment by dfarr** [comment]
Looks great @favalos! The only other change I would recommend would be consolidating the t_api Kind to a single kind for these requests called `CompletePromise`. This is a lot of copypasta work so let me know if you want me to do that and I can add a quick commit.

**Comment by favalos** [comment]
I can work on that tonight, no worries! 

Thanks!

**Comment by favalos** [comment]
@dfarr

Done! Let me know any other suggestions.

Thanks!

File: internal/app/subsystems/api/http/promise.go [hunk 1]
```diff
@@ -122,17 +121,11 @@ func (s *server) completePromise(c *gin.Context) {
...
-	switch strings.ToUpper(body.State) {
-	case promise.Resolved.String():
-		resp, err = s.service.ResolvePromise(id, &header, body)
-	case promise.Rejected.String():
-		resp, err = s.service.RejectPromise(id, &header, body)
-	case promise.Canceled.String():
-		resp, err = s.service.CancelPromise(id, &header, body)
-	default:
+	if !body.State.In(promise.Resolved | promise.Rejected | promise.Canceled) {
 		c.JSON(http.StatusBadRequest, api.HandleValidationError(errors.New("invalid state")))
```
Comment (dfarr) [inline internal/app/subsystems/api/http/promise.go:125]: We just need a return after this line, sorry I didn't see that earlier

**Comment by dfarr** [comment]
Awesome thank you so much @favalos 🎉🎉 🎉 

We just need one return statement and I will merge this PR :)

File: internal/app/subsystems/api/http/promise.go (hunk 1 above)
Comment (favalos) [reply internal/app/subsystems/api/http/promise.go:125]: I added the return after this line, thanks!

**Pull request merged**